### ISSUE TITLE
Reciprocal overlap

### ISSUE THREAD

**Issue by Phillip-a-richmond** [fritzsedlazeck/SURVIVOR#14]
Hello,
First off, I love the tool and it is very useful to our SV analysis for merging call sets between individuals. 

One question with regards to Command #5: 
Is there some way to control reciprocal overlap? I know there is a parameter that controls "within ## BP of each of each other" for merging SVs, but I'm wondering how the logic behind it works.
From your paper:

"Two SVs were defined as overlapping if they occur on the same chromosome, their start and stop coordinates were within 1 kb,"

Is it as such:

SV1: chr1 1000 3000
SV2: chr1 1500 3500

Do you then compute: ( | SV1_start - SV2_start | + | SV1_end - SV2_end | ) ≤ 1000bp
( | 1000 - 1500 | + | 3000 - 3500 | ) = (500 + 500) ≥ 1000, so you count this as the same variant?

Or is it:  | SV1_start - SV2_start | ≤ 1000 AND  | SV1_end - SV2_end | ≤ 1000, then you count this variant. 

Thanks for your help and this awesome tool!

-Phil Richmond @Phillip-A-Richmond

**Comment by fritzsedlazeck** [comment]
Hi Phil,

Sorry for the confusion and thanks for reaching out. Its the second. So both start diff <1000 and end diff < 1000. 

The logic behind is that I saw in populations that we have sometimes polymorphic DUPs in terms of lengths. E.g. we observed a DUP shared in multiple samples that incorporated gene A,B,C but then in other samples a DUP that amplified genes B,C, D,E. So I did want to keep these separated. 

Let me know if you have other questions and as you know 1k is just a parameter. Turns out to work most of the time. In our studies we saw that most of the SVs are within 200p. However, there are sometimes scenarios where 1kp is required.

Cheers
Fritz

**Comment by Phillip-a-richmond** [comment]
Thanks @fritzsedlazeck!
I'm working with this in Human genomes for rare genetic disorders, so I think I may merge variants at different size thresholds, e.g. 
variants 50-200 min distance 50bp,
variants 200-1Kb with min distance 100bp, 
variants 1kb-100kb min distance 500bp, and  
variants 100kb+  min distance 1kb. 

The parameters of min and max size within option #5 should make this trivial, but for post-overlap concatenation do you think a simple cat of the output VCFs (fixing headers of course) would be sufficient? 

Thanks,
Phil

**Comment by fritzsedlazeck** [comment]
Yes that should work.
I have been running it over 22600 human genome samples and kept it constant at 1kp.
If there are variants merged within a sample it records that in the genotype file. 

Let me know how it goes. 
Thanks
Fritz

**Comment by fritzsedlazeck** [comment]
Oh just one more thing. I have been thinking about this and in theory (although I dont have any proof) there should not be a difference in the accuracy of the breakpoints depending on the size of the SVs... 

That might have been true for CNV callers, but for SVs callers that is rather unlikely. However, I would love to be proven wrong :).

**Comment by Phillip-a-richmond** [comment]
Fritz that's not the issue with reciprocal overlap that I was discussing. What I need to avoid is the following situation for smaller CNVs present in two different individuals:
Assume GeneA runs from chr1:15500-18000, and the first codon is from 15500-15600

GeneA_firstCodon: chr1: 15,500-15,600
Now assume we have two SVs, SV1 has no effect, it doesn't hit the gene:

SV_1: chr1 15,000-15,100 DEL
Let's say that SV_1 is present in 30/100 individuals, it's a polymorphism.

Now, let's say that SV2 takes out the first codon of that gene:
SV_2: chr1 15,450-15,650 DEL
This DEL is only present in 1/100 individuals and could be interesting. 

If I use the 1Kb threshold for these SVs, it would merge both of these into the same variant, and when I build the in-house database it will say that SVs from 15,000-15,650 are polymorphic, appearing 31/100 times.

Does this make sense? I'm less worried about consensus calling from different callers on a single individual, and more concerned about considering different variants the same across individuals. By adjusting your distance parameter for the smaller variants, it will fix this issue of completely different variants being merged into a single one. 

Please let me know if something I said or assumed above is incorrect about how SURVIVOR is working for unionizing call-sets between individuals. 

Cheers,
Phil

**Comment by fritzsedlazeck** [comment]
Nope your are right that can happen. 

What SURVIVOR will do in that situation is to report both DEL in the sample column, but count it only once for that sample. 

So makes total sense, I am just curious :). 
Thanks
Fritz

**Comment by Phillip-a-richmond** [comment]
I have 2 questions about SURVIVOR:
1) What does this parameter under option 5 do:
./SURVIVOR 5 

"Estimate distance based on the size of SV"

2) I want to accomplish this:

variants 50-200 min distance 50bp,
variants 200-1Kb with min distance 100bp,
variants 1kb-100kb min distance 500bp, and
variants 100kb+ min distance 1kb.

But there is no max size for this option. Would it be hard to incorporate a max size option into the code?

**Comment by fritzsedlazeck** [comment]
Ad 1 : 
It is a beta version and integrates the function that is implemented in sniffles. You still have a max_distance threshold, but that gets smaller for smaller variations. 

Ad2: 
you mean a max SV length option? I could do that but it will take a bit. I am just about to head out to two conferences and I am quite packed full of work...
Cheers
Fritz

**Comment by fritzsedlazeck** [comment]
Hi Philip,
I just uploaded a new version (just code yet). I changed the interface a bit so watch out.

I included a filtering option where you can extract SVs of certain size ranges. With that you should be able to adjust the merging step. 
Let me know if you have any problems or questions.
Thanks and sorry it took a while.
Fritz

**Comment by Phillip-a-richmond** [comment]
Thanks Fritz,
I'll give it a go and let you know how it runs.

Also, I noticed that in the VCF-->bedpe step, there are discrepancies between the numbering of variants. The same number of variants gets translated, but there are cases where the bedpe SURVIVOR call number differs from the VCF call number. This occurs ~5000 times in a variant vcf file with ~150,000 total variants, and I haven't been able to narrow it down to any specific effect. 

The reason I care about this, is ideally I would have a bedpe file that contained the SUPP=##; and SUPP_VEC=#######..n# annotations. I was using the variant id# to connect between VCF and bedpe in order to parse out these annotations and add them to the bedpe file. After I re-test I can provide a few cases where this is an issue. 

I'll try the new version and let you know how it runs. Thanks for implementing this even though you are busy with conferences.

Talk soon,
Phil

**Comment by fritzsedlazeck** [comment]
Thanks. It might has something todo with TRA events or so, where the size of the event is not clear. I will check it out today. 

Yeah just about the board the next plane... 

Thanks
Fritz

**Comment by Phillip-a-richmond** [comment]
I though that too but it's not the case. I can attach some files here from analysis on open source Platinum Genomes.

I'm attaching 5 files:

1) My shell script where I call SURVIVOR so you can see the commands (SURVIVOR_Commands.sh)
2) The combined VCF file (PlatinumCombinedMetaSVSURVIVOR.vcf)
3) The translated BEDPE file (PlatinumCombinedMetaSVSURVIVOR.bedpe)
4) A python script where I try to add the annotations from VCF to BEDPE (AddFreqToBedpe.py)
5) The variants where issues with naming that occur.  (IssuesWithNaming.txt)

[IssuesWithNaming.txt](https://github.com/fritzsedlazeck/SURVIVOR/files/1679159/IssuesWithNaming.txt)
[SURVIVOR_Commands.sh.txt](https://github.com/fritzsedlazeck/SURVIVOR/files/1679163/SURVIVOR_Commands.sh.txt)
[PlatinumCombinedMetaSVSURVIVOR.vcf.txt](https://github.com/fritzsedlazeck/SURVIVOR/files/1679165/PlatinumCombinedMetaSVSURVIVOR.vcf.txt)
[PlatinumCombinedMetaSVSURVIVOR.bedpe.txt](https://github.com/fritzsedlazeck/SURVIVOR/files/1679166/PlatinumCombinedMetaSVSURVIVOR.bedpe.txt)


Some notes: 
+ They aren't TRA events, or INV, or even INS. 
+ I think it may have to do with variants that share the same starting coordinate (but completely different lengths).
+ There is no loss of variants, and it is not due to min or max size (checked that)

**Comment by Phillip-a-richmond** [comment]
Dunno why I can't attach .py files....

[AddFreqToBedpe.py.txt](https://github.com/fritzsedlazeck/SURVIVOR/files/1679170/AddFreqToBedpe.py.txt)

**Comment by fritzsedlazeck** [comment]
I changed a minor detail in the code for that particular function that should not have any effect for that. 
Then I run one of my vcf files and yours PlatinumCombinedMetaSVSURVIVOR.vcf.txt for vcftobed and checked if I get the same number of entries (using grep -vc '#' ). I get the same number of entries... 

I think what is confusing is that the IDs are actually not parsed. I probably should change that....
That could also explain why you have problems if the starts are the same or close by....

I just pushed the new version. Can you check if this resolves it?
Thanks
Fritz

**Comment by fritzsedlazeck** [comment]
Awesome, thanks for reporting back. 

Yes I will make another release next week. I have been focusing on Sniffles, but since that seems to run fine it is time to put in some work into documentations and versioning for SURVIVOR. 

The dynamic size is still disabled actually. In Sniffles its a percentage from the size of the SV depending on the type. For TRA its the maxium distance. I am still thinking of enabling a parameter for a reciprocal overlap since many people like that idea. 

Thanks
Fritz

**Comment by fritzsedlazeck** [comment]
FYI. I just put up a new release v1.0.2 
Thanks for all the feedback.
Fritz

**Comment by Nazeeefa** [comment]
Hi Fritz,

Is parameter for reciprocal overlap enabled in latest version? Sorry, I couldn't find information on it. I appreciate your help and time, thanks!

Cheers,
Nazeefa

**Comment by fritzsedlazeck** [comment]
Hi Nazeefa,
No it isn't. I will likely take it out completely.
Thanks
Fritz

**Comment by Nazeeefa** [comment]
Hi Fritz,

Just want to double check; if you plan to have/exclude overlap parameter in release 1.0.7? (also, when should we expect new release?). Thank you!

Cheers,
Nazeefa

**Comment by fritzsedlazeck** [comment]
Dear Nazeefa,
I have not implemented an reciprocal overlap parameter. In the newest release, I have however extended the overlap parameter to be specified as a ratio of the length of the SV. This might be very close to that. 
I built a new release 22 days ago.

Cheers
Fritz
